### ISSUE TITLE
tmux fails to start in local session

### ISSUE THREAD

**Issue by kasper93** [microsoft/terminal#5132]
# Environment

Windows 10.0.18363.0
https://github.com/tmux/tmux/commit/f986539e3c1c8757d443ea61c9c155c36aae38d7
WT @ https://github.com/microsoft/terminal/commit/31efd69149c3b23c000c7491f98254356b1eaf34
Fully updated Cygwin as of 2020-03-26
Fully updated MSYS2 as of 2020-03-26

# Steps to reproduce

1. Run tmux

# Expected behavior

tmux works

# Actual behavior

tmux doesn't work
```
open terminal failed: not a terminal
```

### Even more info

tmux works fine in
* Cygwin@Mintty
* MSYS2@Mintty
* wsl
* in ssh session

**Comment by DHowett-MSFT** [comment]
This is almost certainly a Cygwin runtime issue, as it’s failing to detect that the windows console host is attached. MinTTY is somewhat of a red herring- it has a special handshake that it does with Cygwin to assume terminal responsibility. 

How are you launching your Cygwin profile?

**Comment by kasper93** [comment]
For Cygwin I do (it is the same with bash):
```
"commandline": "C:/cygwin/bin/zsh.exe --login -i",
```
For MSYS2 (but it is Cygwin in disguise, so probably the same issue)
```
"commandline": "C:/MSYS/msys2_shell.cmd -defterm -no-start",
```

> MinTTY is somewhat of a red herring- it has a special handshake that it does with Cygwin to assume terminal responsibility.

I understand that. I report here, because it would be nice to make it work in WT. But if you tell me that it need to be fixed on Cygwin side I guess I will redirect my request to them.

I did quick test with ConEmu and it has the same issue,

Ugh, I may have fire to fast with reporting this issue. It seems to be known issue. 
https://github.com/cmderdev/cmder/issues/453
https://github.com/alacritty/alacritty/issues/1687
https://github.com/zeit/hyper/issues/3608

**Comment by zadjii-msft** [comment]
Thanks for following up with all those other issues! Looks like there's a pretty broad problem across terminal emulators on Windows. Do you know if there's actually an upstream bug on cygwin that's tracking this bug? I'd love to resolve this issue as a dup of the upstream one, but I want to make sure the diligence is done ☺️

**Comment by kasper93** [comment]
I searched their ML and there seems to be no report about this specific issue. They don't have public bug tracker except mailing list. I will send them mail and see what they say.

**Comment by DHowett-MSFT** [comment]
Since this looks pretty widespread, I'm gonna close out our copy of it. I'd definitely watch their mailing list, though, if I had a link to an ML archive for this report :smile:

**Comment by kasper93** [comment]
After all I forgot to link here upstream report: https://sourceware.org/pipermail/cygwin/2020-March/244192.html Unfortunately, it went nowhere.

Sometimes you need to fix thing by yourself... maybe when I feel the need for tmux again.

**Comment by changyuheng** [comment]
**Update**
1. Tested both [ConEmu ](https://github.com/Maximus5/ConEmu) and [Cmder](https://github.com/cmderdev/cmder), they don't have this issue. CC @kasper93 
2. Found 2 workarounds for Windows Terminal + MSYS2. CC @sfzhi
   1. Run tmux with this command: `script -c tmux /dev/null`. [Source](https://news.ycombinator.com/item?id=28759834). CC @csdvrx
   2. [windows-alt-sshd-msys2](https://github.com/rkitover/windows-alt-sshd-msys2) CC: @rkitover

**Comment by PrashanthaTP** [comment]
Thank you so much @kasper93  @changyuheng  @sfzhi for the research done and workarounds.
(I will always be thankful to windows terminal team).

TLDR : `script.exe` actually comes with `util-linux` package (msys2)

it may be already known, but commenting here just to save time for someone like me who is
searching for `script`.

---

I was wandering around the internet for months to get [`tmux`](https://github.com/tmux/tmux)
working in windows terminal (without wsl)

For anyone who wants to get `windows terminal`+ `git bash`+`tmux` without installing full package of cygwin or msys, we can do so by getting just `script` binary from msys packages (check once in the msys package page to verify if you already have its dependencies) and putting it in git user bin directory.

References : 

`script.exe` : [packages.msys2.org/package/util-linux](https://packages.msys2.org/package/util-linux?repo=msys&variant=x86_64) 
`scrip.exe is actually in util-linux package` : https://github.com/msys2/MINGW-packages/issues/4497

**Comment by 19945600523** [comment]
改为中文看不懂

**Comment by ducklin5** [comment]
If you are using the `script` `screen` or `dtach` workaround:

You'll need to export a couple of environmental variables in your `.bashrc` or `.zshrc` files to get your cool fonts working:
```bash
export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8
```

**Comment by rkitover** [comment]
Is anyone following up on fixing this in `tmux` itself?

**Comment by dgasaway** [comment]
>     1. Tested both [ConEmu ](https://github.com/Maximus5/ConEmu) and [Cmder](https://github.com/cmderdev/cmder), they do **NOT** have this issue. CC @kasper93

@changyuheng The statement that you tested ConEmu is pretty vague.  I can say that if I try to run cygwin tmux from cmd.exe or cygwin bash in ConEmu, I get the same error as Windows Terminal.  If I host a mintty session in ConEmu, I can run tmux, but then I'm running a different terminal emulator that is known to work.

**Comment by martin-braun** [comment]
The thing is still not fixed, wow.

**Comment by ms-jpq** [comment]
Running `tmux` under `script` seems to be unacceptably slow, especially over SSH.

**Comment by lonelam** [comment]
Hi I just inspected the problem, I checked tmux source code and it seems that the key problem is tmux's tty detection failed on Windows Terminal, while the detection on mintty successed:
https://github.com/tmux/tmux/blob/master/tty.c#L107

So maybe the problem is, on windows, the cmd.exe or terminal.exe indeed not create a tty file to let the tmux to operate, and they are actually not a real terminal according to the following explanation:
https://unix.stackexchange.com/questions/4126/what-is-the-exact-difference-between-a-terminal-a-shell-a-tty-and-a-con

Maybe modifying msys2 could change this behavior? Or just not to use windows for development.

**Comment by rkitover** [comment]
Hi, can anyone interested in this problem please try a tmux branch I'm working on, the police is not letting me test it.

It is the `cygwin-support` branch in my tmux fork here:

https://github.com/rkitover/tmux/tree/cygwin-support

To build do this in an MSYS shell on MSYS2:

```bash
git clone git@github.com:rkitover/tmux
cd tmux
git checkout cygwin-support
pacman -S --noconfirm gcc make bison autoconf automake libevent-devel ncurses-devel pkgconf
sh autogen.sh
./configure --prefix=/usr/local
make -j8
```
, this will create a `tmux.exe` in this directory.
